src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1428 @@
+                SetOfOrCases set, PooledHashSet<LabelSymbol> labelsToIgnore, SyntaxNode nodeSyntax, ref TemporaryArray<StateForCase> casesBuilder)
+            {
+                for (int i = 0; i < patternIndex; i++)
+                {
+                    casesBuilder.Add(previousCases[i]);
+                }
+
+                int index = patternIndex;
+                foreach ((BoundPattern pattern, SyntaxNode? syntax) in set.Cases)
+                {
+                    var label = new GeneratedLabelSymbol("orCase");
+                    SyntaxNode? diagSyntax = syntax;
+                    if (diagSyntax is null)
+                    {
+                        labelsToIgnore.Add(label);
+                        diagSyntax = nodeSyntax;
+                    }
+
+                    Debug.Assert(diagSyntax is not null);
+                    casesBuilder.Add(builder.MakeTestsForPattern(++index, diagSyntax, rootIdentifier, pattern, whenClause: null, label: label));
+                }
+            }
+
+            // We need to reduce the break introduced by flagging redundant patterns
+            // and we never want to affect people who express their patterns thoroughly (but correctly)
+            // such as `switch { < 0 => -1, 0 => 0, > 0 => 1 }`
+            // So we're only reporting a warning for situations that syntactically look hazardous
+            // At the moment, we're only interested in patterns involved in a `not ... or ...` pattern
+
+            // If the pattern is on the right of an `or` pattern, we walk up and
+            // if any of the preceding patterns is a `not` pattern we've detected the error-prone not/or situation.
+            // For example: `not A or <redundant>` or `(A or not B) or <redundant>` or `not B or (A or <redundant>)`
+            static bool detectNotOrPattern(SyntaxNode syntax)
+            {
+start:
+                syntax = unwrap(syntax);
+
+                if (isRightOfOrPattern(syntax, out var binary)
+                    && detectNotPatternOnLeftOfOrPattern(binary))
+                {
+                    return true;
+                }
+
+                Debug.Assert(syntax.Parent is not null);
+                if (isLeftOfOrPattern(syntax, out _)
+                    && isRightOfOrPattern(unwrap(syntax.Parent), out BinaryPatternSyntax? parentBinary)
+                    && detectNotPatternOnLeftOfOrPattern(parentBinary))
+                {
+                    return true;
+                }
+
+                // If the syntax is the whole sub-pattern, we walk up to the recursive pattern.
+                // For example: `not A or { Prop: <redundant> }`
+                if (syntax.Parent is SubpatternSyntax subpatternSyntax
+                    && subpatternSyntax.Parent is (PropertyPatternClauseSyntax or PositionalPatternClauseSyntax) and var patternClause
```

Comment:
>&& subpatternSyntax.Parent is (PropertyPatternClauseSyntax or PositionalPatternClauseSyntax) and var patternClause [](http://example.com/codeflow?start=20&length=114)

It is not obvious why ITuple and list patterns are excluded.

Reply:
I see that we added handling for list patterns, but not for ITuple patterns. Is this intentional?

Reply:
I guess ITuple pattern doesn't have a special syntactic form, it is represented as ```RecursivePatternSyntax```, and. therefore` is handled as well. Is this correct?

Reply:
That's correct. I already had a test for ITuple but missed list pattern